rfcs/51-animation-composition.md

```diff
@@ -0,0 +1,450 @@
+`AnimationGraph`s and corresponding `BoneId`s the entity is bound to.
+
+`BoneBindings` allows us to safely mutate all animated entities in parallel. This
+still requires use of `RefectComponent::reflect_unchecked_mut` or
+`Query::get_unchecked`, but we can be assured that there is no unsound mutable
+aliasing as each entity is only accessed by one thread at any given time.
+
+`BTreeMap`s are used internally as they're typically more iteration friendly
+than `HashMap`. Iteration on `BTreeMap` is `O(size)`, not `O(capacity)` like
+HashMaps, and is array oriented which tends to be much more cache friendly than
+the hash-based scan. It also provides a consistent iteration order for the
+purposes of determinism.
+
+Rough psuedo-code for building `BoneBindings`:
+
+```rust
+fn build_bone_bindings(
+  graphs: Query<(Entity, &AnimationGraph), Changed<AnimationGraph>>,
+  queue: Local<ThreadLocal<...>>, // Parallel queue
+  mut bindings: ResMut<BoneBindings>,
+) {
+    graphs.par_for_each(|(root, graph)| {
+        for (path, bone_id) in graph.bindings() {
+            if let Ok(target) = NameLookup::find(root, path) {
+                queue.push((target, (root, bone_id)))
+            } else {
+                warn!("Could not find bone at {}", path);
+            }
+        }
+    });
+
+    bindings.clear();
+    update_bindings(&mut bindings, parallel_queue);
+}
+```
+
+### Graph Sampling
+All animated properties on an entity are sampled at the same time.
+Sampling a single value from the current state of the graph has the rough
+following flow:
+
+ - All field paths and tracks for a given entity are fetched for a given bone
+   from the associated graph. Fail if not found.
+ - The curves for the active clips is retrieved from the track's sparse array of
+   curves.
+ - Individual values are sampled from each active curve. Skipping curves with a
+   weight of 0.
+ - `Animatable::blend` is used to blend the sampled values together based on the
+   weights in the influences map.
+
+This approach has a number of performance related benefits:
+
+ - The state of the graph does not need to be evaluated for every sampling.
+ - Curves can be sampled in `ClipId` order based on the influence map. This
+   iteration should be cache friendly.
+ - No allocation is inherently required during sampling.  (`Animatable::blend`
+   may require it for some types).
+ - This process is (mostly) branch free and can be accelerated easily with SIMD
+   compatible `Animatable::blend` implementations.
+
+Rough pseudo-code for sampling values with an exclusive system.
+
+```rust
+fn sample_animators(
+  world: &mut World, // To ensure exclusive access
+  bindings: Res<BoneBindings>,
+) {
+    let world: &World = &world;
+    let type_registry = world.get_resource::<TypeRegistry>().clone();
+    // Uses bevy_tasks::iter::ParallelIterator internally.
+    bindings.par_for_each_bone(move |bone, bone_id, graph| {
```

Comment:
Bikeshedding a bit here: `AnimatedEntityBindings` might be a better name.

The end goal is indeed reflect based "animate anything", and there's multiple in flight PRs to make that both possible and performant. This design is explicitly made to avoid making decisions that would block this end goal. Should this defer to #49 for more details on this or should this be elaborated in more detail the implementation *process* not just strategy? Since it clear we likely are going to go through the statically typed Transform based solution first which will be extended to support the "animate anything" use cases.

Reply:
Yeah the challenge is that RFCs must be "implementable", meaning that logical dependencies in other RFCs should be merged first, and the path for implementation should be clear.

I think we have a number of paths forward, which all work for me:
1. Adapt this RFC to be more generic / use the generic design/terminology as defined in #49. Block merging on #49
2. Add the minimal amount of generic design / terminology to this RFC to unblock making it generic. #49 can move at its own pace and iterate on what is introduced here.
3. Adapt this RFC to be less generic and explicitly only target bone transform animation in a way that is implementable (probably just static transform types).

Reply:
I went with option 3 here, since delivering composition is probably a stronger need right now than generic animation, especially without good editor support.

Reply:
That makes sense to me!